File: LC-Java/src/List/_142_detectCycle.java
package List;

import DataStructure.ListNode;

public class _142_detectCycle {
    boolean falg =false;

    public static void main(String[] args) {

    }

    public ListNode detectCycle(ListNode head) {
        if (!falg) {
            return null;
        }
        if (falg) {
            ListNode p1 = hasCycle(head);
            ListNode p2=head;
            int cnt=0;
            while (p1 != p2) {
                p1=p1.next;
                p2=p2.next;
                cnt++;
            }
            return p1;

        }
        return null;
    }

    private ListNode hasCycle(ListNode head) {
        ListNode fast = head, slow = head;
        while (fast != null||fast.next != null) {
            fast=fast.next;
            slow=slow.next;
            if (fast==slow) {
                falg=true;
                return slow;
            }
        }
        return null;
    }
}

File: ImportantAlg/Alg/02-doublePoint/142-detectCycle.go
package doublepoint

import (
	"alg"
)

type ListNode = alg.ListNode

// 142-找到环入口
// 一遍遍历，不用 hasCycle 函数
func detectCycle(head *ListNode) *ListNode {
	if head == nil || head.Next == nil {
		return nil
	}
	slow, fast := head, head
	for fast != nil && fast.Next != nil {
		fast, slow = fast.Next.Next, slow.Next
		if fast == slow {
			fast = head
			for fast != slow {
				fast, slow = fast.Next, slow.Next
			}
			return slow
		}
	}
	return nil
}
func hasCycle(head *ListNode) (bool, *ListNode) {
	if head == nil || head.Next == nil {
		return false, nil
	}
	fast, slow := head, head
	for fast != nil && fast.Next != nil {
		fast, slow = fast.Next.Next, slow.Next
		if fast == slow {
			return true, fast
		}
	}
	return false, nil
}
